doc/tutorials/advanced-ensemble.md

```diff
@@ -0,0 +1,603 @@
+distributed name resolution of servers, and broadcast facilities for
+connecting nodes in an implicit ensemble wide network.
+:::
+
+## Ensemble Domains
+
+As we all know, naming is one of the hardest problems in Computer
+Sciences.  Names need to be intentional, and convey information about
+the role while on the same time avoiding clashes within the global
+namespaces.
+
+Gerbil implements ensemble domains as a hierarchical
+namespace. Servers have identifiers which are pairs of a symbol (the
+_name_ or id of the server) and another symbol which is the
+_domain_. Names cannot contain slashes while domains are hierarchically
+partitioned with slashes.
+
+The hierarchical structure of the domain namespace serves two purposes:
+- it designates control structure, where we can easily refer to the subensemble
+  that operations under a particular subdomain.
+- it eliminates names clashes with easy conventions; for example a convention
+  for the namespace would be `/project/host-id/application/...` and so on.
+
+In addition, each server in a domain has a primary _role_, which
+defines its purpose and its configuration. Servers can also have
+secondary roles and provide a selector for refering to groups of
+servers spanning multiple domains.
+
+## Structured Ensembles
+
+In a _structured ensemble_, a supervisor controls a subdomain and
+provide necessary services for maintaining the ensemble. It manages
+creation of servers, local server-level supervision, and automatically
+spanws and supervises a registry for the entire subdomain in the
+localhost.
+
+Supervisors can be recursive, whereby a supervisor supervises a nested
+supervisor in a local ensemble or meta-supervises remotely as part of
+an orchestrator. In general we recommend as best practice to use a
+single supervisor for a local subdomain and configure meta-supervision
+through orchestrators.
+
+## Case Study: The Gerbil httpd
+
+A common requirement is to spawn an httpd that can utilize all cores,
+and preferably with process isolation so that a bug can't bring down
+the whole server.
+
+Gerbil comes with a builtin httpd, runnable as `gerbil httpd`. The
+server is quite capable, supports static file serving with a small
+file cache, dynamically loadable handlers, and servlets which are interpreted
+handlers in the server content.
+
+### Using a Standalone httpd
+Here we walk through an example www project and how to serve it with `gerbil httpd server`.
+
+First, let's compile our site's code -- see [src/tutorial/advanced-ensemble](https://github.com/mighty-gerbils/gerbil/tree/master/src/tutorial/advanced-ensemble) for the code.
+```
+1$ pushd site/project
+1$ gerbil build
+1$ popd
+```
+
+Now let's configure the httpd. Specifically:
+- we set `www` as the root directory for the site
+- we enable servlets
+- we set a handler for the code we just build
+
+```
+1$ pushd site
+1$ gerbil httpd -G project/.gerbil config --root www --enable-servlets --handlers '(("/handler" . :demo/handler))'
```

Comment:
"-G" is used here but I can't find it in `gxhttpd/opt.ss` or the base `httpd`. Did I miss it?

Reply:
it is in tools/env.ss